### ISSUE TITLE
issue with ImageMol performance

### ISSUE THREAD

**Issue by SoodabehGhaffari** [HongxinXiang/ImageMol#16]
Hello,
[training_class.out.txt](https://github.com/HongxinXiang/ImageMol/files/13811927/training_class.out.txt)
[train_reg.out.txt](https://github.com/HongxinXiang/ImageMol/files/13811928/train_reg.out.txt)

I am reaching out to seek help regarding the ImageMol package. I have used ImageMol to predict the antibiotic activity of compounds both for continuous and binary data. However, I am facing some issues and I am hoping you or one member of your group could help me in this regard.
For classification, the log file shows that the AUC is 0.725, however, I noticed all the probabilities are close to zero and therefore all the y_pred are zero. I was wondering if you have any suggestions to solve this problem. I am not sure why all the y_pred are zero with this high AUC.

For regression, all the y_score have the same value.

I really appreciate your help in this matter.

For your reference, I am attaching the log files for both classification and regression. Could you see if there is anything unusual in the log files?

Best Regards
Soo

**Comment by HongxinXiang** [comment]
Hi, Soo

Thank you first for your attention to our papers.

In some cases, AUC and accuracy may conflict. This is because the accuracy is calculated based on the default cutoff value (such as 0.5), while the AUC is calculated based on all possible cutoff values, which should be more robust.

I am not sure whether your sample has class imbalance. If so, you can try to increase the weight of the minority class.

You can also provide me with the predicted probabilities and corresponding ground truth pickle or numpy files so that I can do better analysis.

**Comment by SoodabehGhaffari** [comment]
Thank you for the prompt reply. 
Yes, our training and test data are so imbalanced (around 1% positive). Could you guide how to increase the weight of the minority class and what value should I set?

Here are the ground truth and the predicted probabilities for the test data:

**Comment by SoodabehGhaffari** [comment]
[df_test.csv.txt](https://github.com/HongxinXiang/ImageMol/files/13820534/df_test.csv.txt)
[df_pro_imagemol_classification.csv](https://github.com/HongxinXiang/ImageMol/files/13820536/df_pro_imagemol_classification.csv)

**Comment by HongxinXiang** [comment]
The following figure is the AUC curve I drew based on the provided `df_pro_imagemol_classification.csv` and `df_test.csv.txt`:
<img width="328" alt="image" src="https://github.com/HongxinXiang/ImageMol/assets/95841534/b045bbd2-e233-4a2b-873d-5e48e295ea07">
Given that your sample is unbalanced, there are two approaches to consider:
1. Find the best classification threshold based on the AUC curve:
```
def find_optimal_cutoff(tpr, fpr, threshold):
    y = tpr - fpr
    index = np.argmax(y)
    optimal_threshold = threshold[index]
    point = [fpr[index], tpr[index]]
    return optimal_threshold, point
fpr, tpr, threshold = roc_curve(y_true, y_pro) 
find_optimal_cutoff(tpr, fpr, threshold)
# my output is (0.001, [0.22790055248618785, 0.5833333333333334])
```
so, I use 0.001 as classification threshhold:
```
a = np.array(y_pro).copy()
a[a>0.001] = 1
a[a<0.001] = 0
(a == y_true).sum() / 2208
```
I can get 0.7685688405797102 accuracy. 

2. Add the weight parameter in [BCEWithLogitsLoss](https://github.com/HongxinXiang/ImageMol/blob/master/finetune.py#L201C24-L201C41). You can look the [docs](https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html). it's very easy to get started.

Anyway, in cases of extremely imbalanced samples, I recommend reporting the AUC metric since it is a more comprehensive metric and better suited to imbalanced data.

In addition, I'm not sure that why the predicted values for the regression model for the test data are the same. But I'm guessing that your regression labels may have a large gap, causing the model to collapse during training. I suggest you can use some normalization method on the labels.

**Comment by SoodabehGhaffari** [comment]
Thank you for the detailed response. I really appreciate your help.
1. Does it make sense to have a so small threshold for the classification model such as 0.001?
2. Regarding the regression model, the training data for the classification and regression is the same with one difference: the continuous label in the regression was converted to binary for the classification. Since the data for classification is imbalanced, I can tell that the labels have a gap. Most of the labels are between zero to 10%. Do you have any suggestions for the normalization method?

Thank you
Best Regards
Soo

**Comment by HongxinXiang** [comment]
Sorry for the late reply.

1. This is an open question and it's hard for me to answer you. In my experience, I would not focus on model accuracy in imbalanced data because it is meaningless. So, I would focus on the AUC since it has nothing to do with the threshold.
2. It is correct.
3.You can try StandardScaler from scikit-learn library.

**Comment by SoodabehGhaffari** [comment]
Hello,
I wanted to give you an update that I tried using the positive weights for the minority class in the loss function as we discussed before, but the issue persists. The values of predicted probability are too small. Do you have any other suggestions or do you think there is a way to fix this issue?

Thanks a lot 
Best Regards
Soo
